### DISCUSSION TITLE
How to flash "Adafruit Feather M4 CAN Express"

### DISCUSSION THREAD

**Discussion by tauonteilchen** [jgressmann/supercan#8] [General]
Hello,

is there a way to flash superCAN firmware to a [Feather M4 CAN Express](https://www.adafruit.com/product/4759) from Adafruit? It is ATSAME51 based but it seems there is not JTAG connection possible.

The "Feather M4 CAN Express" has a [UF2 Bootloader](https://learn.adafruit.com/adafruit-feather-m4-express-atsamd51/uf2-bootloader-details).

**Comment by jgressmann** [comment]
Ha, so Adafruit finally made a board with CAN-FD and USB support. Excellent news.

I am afraid there is no support in SuperCAN for this board at the moment. Having looked at the [schematics](https://github.com/adafruit/Adafruit-Feather-M4-CAN-PCB) this should be very doable.

**Reply by tauonteilchen** [reply]
Ok, so what is to do?
 - supercan_feather-m4-can.h based on  supercan_D5035_01.h?
 - supercan_feather-m4-can.c based on  supercan_D5035_01.c?

I think CAN Bus and LED I/Os must be modified. Feather M4 CAN has a LED at PA23 and a NEO Pixel at PB02
The board supports only one CAN over CAN1 CAN1_TX over PB14 an CAN1_RX over PB15

The schematic shows a TCAN1051HGV transceiver, on my sample board seems only a TCAN1051V so may be only 2Mbit possible

**Comment by jgressmann** [comment]
I added initial support for the board [here](https://github.com/jgressmann/tinyusb/tree/feat/supercan/feather_m4_can_express).
You need to build with `BOARD=feather_m4_express` (files hacked for CAN variant).
Since I don't have a board, the state of the code is that of an educated guess at best. It should get you started, however.

**Reply by tauonteilchen** [reply]
Thank you, may I found a way to connect a [Segger J-Link](https://embeddedcomputing.weebly.com/segger-j-link-with-adafruit-feather-m4.html). I have to solder a bit. I'll try to test it at this weekend.

**Comment by tauonteilchen** [comment]
I could flash the firmware to Feather M4 Express. Driver should installed.
dmesg says:
```
[  982.431543] usb 1-1: new full-speed USB device number 20 using xhci_hcd
[  982.431741] usb 1-1: Device not responding to setup address.
[  982.639995] usb 1-1: Device not responding to setup address.
[  982.847810] usb 1-1: device not accepting address 20, error -71
[  982.847943] usb usb1-port1: unable to enumerate USB device
```

**Comment by jgressmann** [comment]
[Release 1.0.14](https://github.com/jgressmann/supercan/releases/tag/v1.0.14) adds proper support for the board.

**Reply by tauonteilchen** [reply]
Thank you very much.
I flashed the firmware via UF2 Bootloader, just drag and drop. It works like a normal CAN device.
